drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+            },
+        }
+    }
+}
+
+/// Does (component, slot) refer to the currently running Hubris image?
+pub fn is_current_hubris_image(component: RotComponent, slot: SlotId) -> bool {
+    // Safety: We are trusting the linker.
+    flash_range(component, slot).store.start
+        == unsafe { &__this_image } as *const _ as u32
+}
+
+// LPC55 defined image content
+
+/// Image header for the LPC55S6x device as documented in NXP UM11126
+#[repr(C)]
+#[derive(Default, AsBytes, FromBytes)]
+pub struct ImageVectorsLpc55 {
+    initial_sp: u32,                    // 0x00
+    initial_pc: u32,                    // 0x04
+    _vector_table_0: [u32; 6],          // 0x08, 0c, 10, 14, 18, 1c
+    nxp_image_length: u32,              // 0x20
+    nxp_image_type: u32,                // 0x24
+    nxp_offset_to_specific_header: u32, // 0x28
+    _vector_table_1: [u32; 2],          // 0x2c, 0x30
+    nxp_image_executation_address: u32, // 0x32
+                                        // Additional trailing vectors are not
+                                        // interesting here.
+                                        // _vector_table_2[u32; 2]
+}
+
+impl ImageVectorsLpc55 {
+    const IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE: u32 = 4;
+
+    pub fn is_image_type_signed_xip(&self) -> bool {
+        self.nxp_image_type == Self::IMAGE_TYPE_PLAIN_SIGNED_XIP_IMAGE
+    }
 
     // This part aliases flash in two positions that differ in bit 28. To allow
     // for either position to be used in new images, we clear bit 28 in all of
     // the numbers used for comparison below, by ANDing them with this mask:
-    const ADDRMASK: u32 = !(1 << 28);
+    pub fn normalized_initial_pc(&self) -> u32 {
+        const ADDRMASK: u32 = !(1 << 28);
+        self.initial_pc & ADDRMASK
+    }
+
+    // Length of image from offset zero to end of the signature block (without padding)
+    // a.k.a. ImageVectorsLpc55.nxp_image_length
+    pub fn image_length(&self) -> Option<u32> {
+        if self.is_image_type_signed_xip() {
+            Some(self.nxp_image_length)
+        } else {
+            None
+        }
+    }
+
+    /// Image length padded to nearest page size.
+    pub fn padded_image_len(&self) -> Option<u32> {
+        round_up_to_flash_page(self.image_length()?)
+    }
+
+    /// Determine the bounds of an image assuming the given flash bank
+    /// addresses.
+    pub fn padded_image_range(&self, exec: &Range<u32>) -> Option<Range<u32>> {
+        let image_start = exec.start;
+        let image_end = exec.start.saturating_add(self.padded_image_len()?);
+        // ImageHeader is optional. Assuming code cannot be earlier than
+        // IMAGE_HEADER_OFFSET.
+        let initial_pc_start = image_start.saturating_add(IMAGE_HEADER_OFFSET);
+        let initial_pc_end =
+            image_start.saturating_add(self.nxp_offset_to_specific_header);
+        let pc = exec.start.saturating_add(self.normalized_initial_pc());
+
+        if !exec.contains(&image_end)
```

Comment:
Because this isn't using `RangeInclusive` and you're adding the length to the start, an image that takes up the entire flash slot won't pass this check.